fix/RFE-358, RFE-359/update eslint and eslint-plugin-jsdoc to fix vulnerabilities

roma-derski opened the pull request:
[RFE-358/RFE-359](https://oat-sa.atlassian.net/wiki/spaces/OAT/pages/2009891385/RFE-358+RFE-359+-+Update+Node+and+NPM)
BREAKING CHANGE
- update eslint and eslint-plugin-jsdoc versions to fix vulnerabilities & peer dependencies issues
- update eslint-plugin-jest version to address compatibility issue for aforementioned update

jsconan commented on package.json:
```diff
@@ -27,18 +27,19 @@
   "publishConfig": {
     "access": "public"
   },
+  "engines": {
+    "node": ">=18.15.0",
+    "npm": ">=9.5.0"
+  },
```
**issue:** This is a shared dependency. We should not force the version here unless the features require it, but I don't think it is the case. Please let it open.

roma-derski replied:
agree, removed

jsconan commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@oat-sa/eslint-config-tao",
-  "version": "1.2.0",
+  "version": "1.2.1",
```
**issue:** The version is handled by the [release tool](https://github.com/oat-sa/tao-extension-release), so no need to update it.

The version itself is not correct as it should be a major change since the base dependency was also upgraded to a major.

roma-derski replied:
got it, reverted

jsconan commented on .npmrc:
```diff
@@ -0,0 +1 @@
+engine-strict=true
```
**issue:** Please don't make a library strict unless it is motivated by a minimal support for the required features.

roma-derski replied:
agree, removed

jsconan reviewed (changes requested):
While it is ok to set the minimal version on an application, it is not on a library. Nothing motivates us to force Node 18 for this library.

The update for ESLint is nice as it fixes issues. However, the version change should be a major, not a patch. This means a commit with a breaking mention.

jsconan commented on package-lock.json:
```diff
@@ -1,1279 +1,2297 @@
   "name": "@oat-sa/eslint-config-tao",
   "version": "1.2.0",
-  "lockfileVersion": 1,
+  "lockfileVersion": 3,
```
**issue:** Please revert also the lock file. For the sake of the update, it needs to be generated using `npm 6`, until we moved all the packages to `npm 7+` otherwise it may break the install using `npm ci`

roma-derski replied:
reverted

oatymart replied:
I still see lockfileVersion 3, and installing with npm 6 causes it to be reverted to 1. I think 1 should be kept.

roma-derski replied:
my bad, reverted to 1

jsconan reviewed (commented):
It looks better, however, it would be better to generate the lock file using Node 14/NPM 6 for the time being. After the migration it should not be an issue, but for now it is better to keep the existing process.

roma-derski commented:
> While it is ok to set the minimal version on an application, it is not on a library. Nothing motivates us to force Node 18 for this library.
> 
> The update for ESLint is nice as it fixes issues. However, the version change should be a major, not a patch. This means a commit with a breaking mention.

added breaking mention

jsconan commented on package.json:
```diff
@@ -32,13 +32,10 @@
     "eslint-plugin-svelte3": "^3.1.1"
   },
   "devDependencies": {
-    "eslint": "^7.22.0"
-  },
-  "peerDependencies": {
-    "eslint": "^7.22.0"
```
**issue:** We need either to keep the peer dependency as this package was meant for configuration and we still need to declare which version is recommended for the consumers. Or we must either move the eslint to transitive dependencies.

roma-derski replied:
indeed. checked peerDependencies for all dependencies involved:

- @babel/eslint-parser: eslint: ^7.5.0 || ^8.0.0
- eslint-plugin-cypress: eslint: >= 3.2.1
- eslint-plugin-es: eslint: >=4.19.1
- eslint-plugin-jest: eslint: >=5
- eslint-plugin-jsdoc: eslint: ^7.0.0 || ^8.0.0
- eslint-plugin-svelte3: eslint: >=6.0.0

so peerDependencies of eslint ^8.0.0 satisfies all of them - updated the code accordingly (before, with the lower eslint-plugin-jsdoc version it was not the case so eslint ^7 was used).

jsconan reviewed (commented):
It looks almost ok. However, there is one last thing I did not see in the previous review, and it needs to be addressed too.

jsconan commented:
Added @oatymart and @btamas for the review.

One remark before merging, the PR mentions there is a breaking change, but no commit enforces it. The release tool will not increase to the correct version, and it would need to force it.

oatymart reviewed (changes requested):
First a couple of admin notes:
- The PR title looks wrong. For the sake of future searches in https://github.com/oat-sa/eslint-config-tao/pulls, please update it. (Branch name doesn't matter so much, it will be deleted after)
- README.md has a changelog at the bottom, I guess you can update it too.

oatymart commented:
I also had an issue trying to integrate this version to a project.
In [tao-deliver-fe](https://github.com/oat-sa/tao-deliver-fe), I deleted the existing `eslint` (v7) and `oat-sa/eslint-config-tao` folders from node_modules, and ran `npm link @oat-sa/eslint-config-tao` to link this package in.
Then:
```
$ npm run lint                                         v14.18.0  [Tue 18, 12:03]

> tao-deliver-fe@11.22.0 lint /Users/martinnicholson/work/nui-lds/tao-deliver-fe
> eslint -c .eslintrc.js 'packages/**/*.{js,svelte}' 'cypress/**/*.js'


Oops! Something went wrong! :(

ESLint: 8.1.0

Error: .eslintrc.js » @oat-sa/eslint-config-tao/svelte:
	Environment key "jest/globals" is unknown
```
Any idea what has caused this?

roma-derski commented:
> First a couple of admin notes:
> 
> * The PR title looks wrong. For the sake of future searches in https://github.com/oat-sa/eslint-config-tao/pulls, please update it. (Branch name doesn't matter so much, it will be deleted after)
> * A link to a Jira ticket would be nice. I'm assuming RFE-359 can be used
> * README.md has a changelog at the bottom, I guess you can update it too.

- updated PR title
- added link to Jira ticket in PR description
- updated changelog in README

roma-derski commented:
> I also had an issue trying to integrate this version to a project. In [tao-deliver-fe](https://github.com/oat-sa/tao-deliver-fe), I deleted the existing `eslint` (v7) and `@oat-sa/eslint-config-tao` folders from node_modules, and ran `npm link @oat-sa/eslint-config-tao` to link this package in. Then:
> 
> ```
> $ npm run lint                                         v14.18.0  [Tue 18, 12:03]
> 
> > tao-deliver-fe@11.22.0 lint /Users/martinnicholson/work/nui-lds/tao-deliver-fe
> > eslint -c .eslintrc.js 'packages/**/*.{js,svelte}' 'cypress/**/*.js'
> 
> 
> Oops! Something went wrong! :(
> 
> ESLint: 8.1.0
> 
> Error: .eslintrc.js » @oat-sa/eslint-config-tao/svelte:
> 	Environment key "jest/globals" is unknown
> ```
> 
> Any idea what has caused this?

reproduced, will investigate and try to fix

roma-derski commented:
> 

this was caused by outdated eslint-plugin-jest package, so pushed update also for this package

oatymart reviewed (approved):
Finally I achieved a correct linting in tao-deliver-fe. So I'm satisfied we can integrate this package successfully in most projects. Good job!

- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
- [ ] ~~Bundled code is present if needed~~
- [ ] ~~Version bump is applied if needed~~
- [x] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x]  Commits messages are meaningful
- [x] Commits are atomic
- [x] ~~Changelog is updated according to changes (if applicable)~~
- [ ] Documentation is updated according to changes (if applicable)
